### PR TITLE
[jit] Fix layernorm ad formula on weight and bias

### PR DESCRIPTION
**Pull request by wanchaol**
Fix the layernorm formula when weight and bias passed in.

### CODE REVIEW

File: torch/csrc/jit/symbolic_script.cpp [hunk 1]
```diff
@@ -708,16 +715,32 @@ const std::vector<std::string> functions = {
...
             def backward(grad_output):
-                if weight is not None:
-                    grad_output = grad_output * torch.t(weight)
-                    weight = grad_output * torch.t(bn_out)
+                if weight is not None and bias is not None:
+                    grad_bn_out = (grad_output * weight)._grad_sum_to_size(bn_out.size())
+                    grad_weight = (grad_output * weight)._grad_sum_to_size(weight.size())
+                    grad_bias = grad_output._grad_sum_to_size(bias.size())
+                elif weight is not None:
+                    grad_bn_out = grad_output._grad_sum_to_size(bn_out.size())
```
Comment (ngimel) [inline torch/csrc/jit/symbolic_script.cpp]: `grad_output * weight`

File: torch/csrc/jit/symbolic_script.cpp [hunk 2]
```diff
@@ -708,16 +715,32 @@ const std::vector<std::string> functions = {
...
+                    grad_weight = (grad_output * bn_out)._grad_sum_to_size(weight.size())
+                    grad_bias = None
+                elif bias is not None:
+                    grad_bn_out = grad_output._grad_sum_to_size(bn_out.size())
```
Comment (ngimel) [inline torch/csrc/jit/symbolic_script.cpp]: _grad_sum_to_size is not required, same for the `grad_bn_out` in the last conditional

File: test/test_jit.py [hunk 3]
```diff
@@ -13148,9 +13150,7 @@ class TestJitGeneratedFunctional(JitTestCase):
     ('batch_norm', (S, S), (non_differentiable(torch.randn(S)), non_differentiable(torch.ones(S)), ),),
     ('instance_norm', (S, S, S), (non_differentiable(torch.zeros(S)), non_differentiable(torch.ones(S))),),
     ('layer_norm', (S, S, S, S), ([5],),),
-    ('layer_norm', (S, S, S, S), ([5], (S,)), 'with_only_weight'),
-    ('layer_norm', (S, S, S, S), ([5], None, (S,)), 'with_only_bias'),
-    ('layer_norm', (S, S, S, S), ([5], (S,), (S,)), 'with_weight_and_bias'),
```
Comment (ailzhang) [inline test/test_jit.py]: Why do we delete tests here?

Comment (wanchaol) [reply test/test_jit.py]: because for `nn.Layernorm`, we basically only calls with no weight and bias, or with both weight and bias, it does not make too much sense to add the tests to test `only_weight` and `only_bias`

File: torch/csrc/jit/symbolic_script.cpp [hunk 4]
```diff
@@ -708,16 +715,32 @@ const std::vector<std::string> functions = {
                 output = bn_out
 
             def backward(grad_output):
-                if weight is not None:
-                    grad_output = grad_output * torch.t(weight)
-                    weight = grad_output * torch.t(bn_out)
+                if weight is not None and bias is not None:
+                    grad_bn_out = (grad_output * weight)._grad_sum_to_size(bn_out.size())
```
[the remark this answers is not in the document]
Comment (wanchaol) [reply torch/csrc/jit/symbolic_script.cpp]: actually why is that? I think `addcmul` is a broadcastable op, so it is necessary for us to reduce the size when bn_out, weight, bias aren't the same shape?

File: test/test_jit.py (hunk 3 above)
Comment (wanchaol) [reply test/test_jit.py]: synced offline, it looks like we also expose nn.functional as public API so there will be cases people pass in only weight and bias, I will bring it back

File: torch/csrc/jit/symbolic_script.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (ngimel) [reply torch/csrc/jit/symbolic_script.cpp]: It broadcasts to the size of bn_out, so the size of grad_bn_out and bn_out will be the same.

[the remark this answers is not in the document]
Comment (wanchaol) [reply torch/csrc/jit/symbolic_script.cpp]: I see, that make sense since weight and bias are in the shape of normalized_shape

File: torch/csrc/jit/symbolic_script.cpp [hunk 5]
```diff
@@ -708,16 +715,32 @@ const std::vector<std::string> functions = {
                 output = bn_out
 
             def backward(grad_output):
-                if weight is not None:
-                    grad_output = grad_output * torch.t(weight)
-                    weight = grad_output * torch.t(bn_out)
+                if weight is not None and bias is not None:
+                    grad_bn_out = grad_output * weight
+                    grad_weight = grad_output * bn_out
```
Comment (ngimel) [inline torch/csrc/jit/symbolic_script.cpp]: you do need sum_to_size here and for other grad_weight terms (sum_to_size is only unnecessary for grad_bn_out). Also, sum_to_size is needed for grad_bias terms. Why aren't tests catching this?

Comment (wanchaol) [reply torch/csrc/jit/symbolic_script.cpp]: hmmm yeah you are right, thanks for pointing out. I think the reason that the tests did not catch those is because we only check `grad_input` in our testing infra, I will see if I can change that.

**Comment by ailzhang** [comment]
Sorry this PR also need a rebase to include the test_jit change( and properly enabling AD check for `layernorm`.

**Review by ailzhang — APPROVED** [review]
